identity/src/main/java/com/android/identity/credential/NameSpacedData.kt

```diff
@@ -0,0 +1,386 @@
+        val mapBuilder: MapBuilder<CborBuilder> = builder.addMap()
+        for (namespaceName in map.keys) {
+            val innerMapBuilder: MapBuilder<MapBuilder<CborBuilder>> =
+                mapBuilder.putMap(namespaceName)
+            val namespace = map[namespaceName]!!
+            for (dataElementName in namespace.keys) {
+                val dataElementValue = namespace[dataElementName]
+                val taggedBstr = ByteString(dataElementValue)
+                taggedBstr.tag = Tag(24)
+                innerMapBuilder.put(UnicodeString(dataElementName), taggedBstr)
+            }
+        }
+        return builder.build().get(0)
+    }
+
+    /**
+     * Encodes the given [NameSpacedData] as CBOR.
+     *
+     * The encoding uses the following CDDL:
+     * ```
+     * NameSpacedCbor = {
+     * + NameSpaceName =&gt; DataElements
+     * }
+     *
+     * NameSpaceName = tstr
+     *
+     * DataElements = [
+     * + DataElementName =&gt; DataElementValueBytes
+     * ]
+     *
+     * DataElementName = tstr
+     * DataElementValue = any
+     * DataElementValueBytes = #6.24(bstr .cbor DataElementValue)
+     * ```
+     *
+     * Here's an example of CBOR conforming to the above CDDL printed in diagnostic form:
+     * ```
+     * {
+     *   "org.iso.18013.5.1" : {
+     *     "given_name" : 24(&lt;&lt; "Erika" &gt;&gt;),
+     *     "family_name" : 24(&lt;&lt; "Mustermann" &gt;&gt;),
+     *   },
+     *   "org.iso.18013.5.1.aamva" : {
+     *     "organ_donor" : 24(&lt;&lt; 1 &gt;&gt;)
+     *   }
+     * }
+     * ```
+     *
+     * Name spaces and data elements will be in the order they were inserted using
+     * at construction time, either using the [.Builder] or [.fromEncodedCbor].
+     *
+     * @return the bytes of the encoding describe above.
+     */
+    fun encodeAsCbor(): ByteArray {
+        return Util.cborEncode(toCbor())
+    }
+
+    /**
+     * A builder for [NameSpacedData].
+     */
+    class Builder {
+        var map = LinkedHashMap<String, LinkedHashMap<String, ByteArray>>()
+
+        /**
+         * Adds a raw CBOR value to the builder.
+         *
+         * For performance-reasons the passed in value isn't validated and it's the
+         * responsibility of the application to perform this check if for example operating
+         * on untrusted data.
+         *
+         * @param nameSpaceName the name space name.
+         * @param dataElementName the data element name.
+         * @param value the bytes of the CBOR.
+         * @return the builder.
+         */
+        fun putEntry(
+            nameSpaceName: String,
+            dataElementName: String,
+            value: ByteArray
+        ): Builder {
+            var innerMap = map[nameSpaceName]
+            if (innerMap == null) {
+                innerMap = LinkedHashMap()
+                map[nameSpaceName] = innerMap
+            }
+            innerMap[dataElementName] = value
+            return this
+        }
```

Comment:
We can use `putIfAbsent` that only adds to the map if the specified key does not exist

```
fun putEntry(
            nameSpaceName: String,
            dataElementName: String,
            value: ByteArray
        ): Builder {
            map.putIfAbsent(nameSpaceName, mutableMapOf())
            map[nameSpaceName]?.run {
                this[dataElementName] = value
            }
            return this
        }
```